internal/watcher/watcher.go

```diff
@@ -231,3 +333,168 @@ func NewDifferential(
+		}
+		if os.SameFile(sfi, dfi) {
+			return
+		}
+	}
+
+	err = copyFileContents(src, dst)
+	return
+}
+
+// copyFileContents copies the contents of the file named src to the file named
+// by dst. The file will be created if it does not already exist. If the
+// destination file exists, all it's contents will be replaced by the contents
+// of the source file.
+func copyFileContents(src, dst string) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+	if _, err = io.Copy(out, in); err != nil {
+		return err
+	}
+	err = out.Sync()
+	return err
+}
+
+func after(value string, a string) string {
+	// Get substring after a string.
+	pos := strings.LastIndex(value, a)
+	if pos == -1 {
+		return ""
+	}
+	adjustedPos := pos + len(a)
+	if adjustedPos >= len(value) {
+		return ""
+	}
+	return value[adjustedPos:]
+}
+
+// WatchLargeFileFn is function to read large file then puts the contents of the file into a temporary directory
+// and saves the file path to previousContents
+func (d Differential) WatchLargeFileFn(info fs.FileInfo, path string) error {
+
+	err := CopyFile(path, TemporaryDirectory+"/"+info.Name())
+	if err != nil {
+		return fmt.Errorf("failed to copy file  %w", err)
+	}
+
+	// find diff
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("ioutil.ReadFile: %w", err)
+	}
+
+	checksum := md5.Sum(contents) //nolint:gosec
+	if checksum == d.previousChecksum[path] {
+		return nil
+	}
+
+	diff := cmp.Diff(string(d.previousContents[path]), string(contents))
+
+	// set new previous content
+	contentName := "FilePath##" + TemporaryDirectory + "/" + info.Name()
+	d.previousContents[path] = []byte(contentName)
+
+	if err := d.sendDiff(path, diff); err != nil {
+		return fmt.Errorf("d.sendDiff: %w", err)
+	}
+
+	log.Printf("Sent the diff (size = %d)", len(diff))
+
+	return nil
+}
+
+func (d Differential) readFileAndFindDiff(path, srcFile string) (err error) {
+
+	f1, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("error openning watched file : %w", err)
+	}
+	f2, err := os.Open(srcFile)
+	if err != nil {
+		return fmt.Errorf("error openning temp file: %w", err)
+	}
+	defer f1.Close()
+	defer f2.Close()
+
+	r1 := bufio.NewReader(f1)
+	r2 := bufio.NewReader(f2)
+
+	eof1 := false
+	eof2 := false
+
+	for {
+		buf1 := make([]byte, MaxReadSize)
+		buf2 := make([]byte, MaxReadSize)
+		// read first file
+		n, err := r1.Read(buf1)
+		if err == io.EOF {
+			eof1 = true
+		} else if err != nil {
+			return fmt.Errorf("error read chunkz watched file: %w", err)
+		}
+		content1 := buf1[:n]
+
+		// read second file
+		n, err = r2.Read(buf2)
+		if err == io.EOF {
+			eof2 = true
+		} else if err != nil {
+			return fmt.Errorf("error read chunkz watched file: %w", err)
+		}
+		content2 := buf2[:n]
```

Comment:
By completely reading both files into memory a lot of memory is allocated just to be de-allocated shortly after as both files are the same in 99,999% of the cases. Is it suitable to compare the md5 of the changed files first against the md5 of the possible changed one?

Reply:
@hanzei is this what you are referring to https://codereview.stackexchange.com/questions/109942/comparing-md5-sum-to-text-from-file

Reply:
I don't quite get your question. The link you posted shows an example implementation of a similar issue, but this one here is simpler. Both buffers are bytes slices.

Reply:
@hanzei  I can only think of two days to check the difference 
1. we compare the files the first time and store the checksum inside a store and then compare it or update it depending on the use case in the logic. 
2. the implementation here where we compare the files. 

Do you have any other ways that you think would be more performant ? or shall we go for 1 above ?

Reply:
@hanzei @maisnamrajusingh this code has been removed